Sources/Utility/ArgumentParser.swift

```diff
@@ -447,3 +447,166 @@ public final class ArgumentParser: ArgumentParserProtocol {
         }
     }
 }
+
+/// A protocol representing an option structure, which can be
+/// bounded in the BoundParser.
+public protocol BoundParserOption {
+    init()
+}
+
+/// BoundParser is ArgumentParser bounded to an option, which provides
+/// API helpers to fill values in the bounded option.
+///
+/// Example:
+///         struct Options: BoundParserOption {
+///             var verbose = false
+///         }
+///         let parser = BoundParser<Options>(usage: "Sample", overview: "Sample")
+///         parser.add(option: "--verbose", kind: Bool.self) { $0.verbose = $1 }
+///         let options = try parser.parse(["--verbose"])
+public final class BoundParser<OptionType: BoundParserOption> {
+    /// The underlying parser instance.
+    private let parser: ArgumentParser
+
+    /// The result of the parsing.
+    private var result: ArgumentParser.Result!
+
+    /// The bounded option.
+    private var options: OptionType
+
+    /// This array holds the closures which should be executed post parsing.
+    /// Calling the closures will fill up the options.
+    private var bodies = [() -> Void]()
+
+    /// Create a bounded parser.
+    public init(commandName: String? = nil, usage: String, overview: String) {
+        parser = ArgumentParser(commandName: commandName, usage: usage, overview: overview)
+        options = OptionType()
+    }
+
+    /// Add an option argument.
+    @discardableResult
+    public func add<T: ArgumentKind>(
+        option: String,
+        shortName: String? = nil,
+        kind: T.Type,
+        usage: String? = nil,
+        body: ((inout OptionType, T) -> Void)? = nil
+    ) -> OptionArgument<T> {
+
+        let arg = parser.add(option: option, shortName: shortName, kind: kind, usage: usage)
```

Comment:
Maybe it would be better for client to invoke the add, and pass the result in? That would avoid needing to duplicate all the signatures.